### ISSUE TITLE
Components list violates CodeList definition

### ISSUE THREAD

**Issue by phackstock** [IAMconsortium/nomenclature#135]
As I was trying to get the new component check tests working on #133 I found a somewhat sneaky error/inconsistency in the way that `CodeList` is using `CodeList` recursively for specifying a list of components as a dictionary. 

## Description

As it stands, in `cast_variable_components_args` (https://github.com/IAMconsortium/nomenclature/blob/main/nomenclature/codelist.py#L104), if we have a list of different aggregation dimensions, we cast them into a `CodeList`:
```python
...
# translate a list of single-key dictionaries to a simple dictionary
            for (name, attrs) in items:
                if not all([isstr(i) for i in attrs["components"]]):
                    values["mapping"][name]["components"] = CodeList(
                        name="components", mapping=attrs["components"]
                    )
```

This means that the value for the "components" attribute in a "variable" codelist can be of type `CodeList`.
This, however clashes with the type definition in the `CodeList` header for the `mapping` attribute (https://github.com/IAMconsortium/nomenclature/blob/main/nomenclature/codelist.py#L53):

```python
 mapping: Union[List, Dict[str, Dict[str, Union[str, float, int, list, None]]]] = {}
```
in the final `Union` which specifies what the final data types are allowed to be we would need an additional `CodeList` to make it work.

The reason why nothing broke despite `pydantic` enforcing types is because are checked **before** the validators are run (except when `pre=True`). The validators are then run under the assumption that they don't recast the type of the variable but only check if it conforms to whatever checks there are in place. 
As I found out today, however, there's nothing stopping you from changing the data type and therefore rendering the entire typing checking pointless. 
A minimum example to illustrate:

```python
from pydantic import BaseModel, root_validator

class A(BaseModel):

    a: int

    @root_validator()
    def change_a(cls, values):
        # convert to list 
        values["a"] = [values["a"]]
        return values

print(A(a=10).a) # prints out [10] instead of the expected 10!
```

This is the exact same thing that is currently happening in the `CodeList` model.
In my opinion this is bad because now `CodeList.variable["some variable"]["components"]` can not only contain a list or a dictionary as the definition would suggest but actually also another `CodeList`.

## Proposed fixes

I would see two fixes for this:
1. Add `CodeList` to the `Union` of allowed types in the final level of the `mapping` attribute. 
2. Change it so that `cast_variable_components_args` does not cast to a `CodeList` but rather directly to a dictionary as the original docstring in `cast_variable_components_args` suggests.

My preference would be option 2 because in option 1 we are opening the door for a recursive `CodeList` structure. This might be desired at some point in the future but I always have troubles with this kind of complexity. 
Option 2 on the other hand I find more simple to visualize and it should be good enough for our purposes. That is of course, unless, we want some kind of validation which `CodeList` conveniently provides for us. I've though about it quickly but apart from making sure that the keys in `"components"` are unique I could not think of anything that would not be covered by our yaml/json schema specification already. That being said if it's only about ensuring that the keys are unique I would just write a quick check before casting to a dict instead of using a whole new instance of `CodeList`.

In any case I would suggest that it might be a good idea to move any kind of re-casting for types to a validator (`root_validator` or normal `validator`) with `pre=True`. All other validators are then only allowed to verify the integrity of the data and not change it. This way we would ensure that the types specified in the class definition of `CodeList` are the types an instance of `CodeList` actually ends up having. 

This one got a bit more lengthy than expected, would love to hear your thoughts @danielhuppmann.

**Comment by danielhuppmann** [comment]
I agree, this was not an elegant solution. The main reason is that I wanted to reuse the casting from a list of single-key-dicts to a dictionary provided by CodeList, including guarding against duplicates and ensuring that there is exactly one key in each entry.

**Comment by phackstock** [comment]
Right, the reasoning totally makes sense, thanks for the quick PR, I'll take a look.
